### ISSUE TITLE
Redis+Sentinel failover failing

### ISSUE THREAD

**Issue by dioguerra** [bitnami/charts#8113]
Hello,

I'm trying the Helm chart from bitnami/redis. I have 3 nodes redis+sentinel running and i am currently trying to force terminate the redis pod (that is master - in my case master is redis-node-0).
I notice that while this pod is in pending state, the redis cluster stops working (with 2 pods up). As the clusters should only require a quorum of 2 nodes, shoudn't the redis keep working?

Error on one of the slaves: 
```
1:S 09 Nov 2021 16:26:55.496 * Connecting to MASTER redis-node-0.redis-headless.kube-system.svc.cluster.local:6379
1:S 09 Nov 2021 16:26:55.499 # Unable to connect to MASTER: Invalid argument
```
Sentinel:
```
1:X 09 Nov 2021 16:12:30.888 # +switch-master mymaster redis-node-1.redis-headless.kube-system.svc.cluster.local 6379 redis-node-0.redis-headless.kube-system.svc.cluster.local 6379
1:X 09 Nov 2021 16:12:30.889 * +slave slave redis-node-2.redis-headless.kube-system.svc.cluster.local:6379 redis-node-2.redis-headless.kube-system.svc.cluster.local 6379 @ mymaster redis-node-0.redis-headless.kube-system.svc.cluster.local 6379
1:X 09 Nov 2021 16:12:30.889 * +slave slave redis-node-1.redis-headless.kube-system.svc.cluster.local:6379 redis-node-1.redis-headless.kube-system.svc.cluster.local 6379 @ mymaster redis-node-0.redis-headless.kube-system.svc.cluster.local 6379
1:X 09 Nov 2021 16:17:09.498 # +sdown sentinel 2a09ba7abbb41ee71e79087310d75f9809c3c815 redis-node-0.redis-headless.kube-system.svc.cluster.local 26379 @ mymaster redis-node-0.redis-headless.kube-system.svc.cluster.local 6379
1:X 09 Nov 2021 16:17:28.857 # +sdown master mymaster redis-node-0.redis-headless.kube-system.svc.cluster.local 6379
```

Instead the slaves get blocked. Once the master pod comes back the sentinels quickly change the master node to a new pod (but only after the master comes back). What is happening?

 Im using helm with the values bellow. 
 Kubectl server 1.21 client 1.22
```
helm upgrade -i redis bitnami/redis --namespace kube-system --version 15.5.4 --values ~/deletemetoo.yaml
```

```yaml
global:
  storageClass: "default"
  redis:
    password: "PASSWORD"

auth:
  sentinel: false

master:
  resources:
    requests:
      cpu: 100m
    limits:
      cpu: 250m
      memory: 1Gi
  persistence:
    size: 10Gi

replica:
  resources:
    requests:
      cpu: 100m
    limits:
      cpu: 250m
      memory: 1Gi
  persistence:
    size: 10Gi

sentinel:
  enabled: true
  # masterSet: mymaster
  automateClusterRecovery: true
  downAfterMilliseconds: 2000
  failoverTimeout: 1000
  resources:
    requests:
      cpu: 100m
    limits:
      cpu: 200m
      memory: 200Mi    
```

**Comment by alvneiayu** [comment]
hi @dioguerra 

Let me investigate a little bit the problem that you are reproducing. I will come back with some feedback as soon as possible.

Thanks a lot

Álvaro

**Comment by movergan** [comment]
Interested to know the results.

**Comment by javsalgar** [comment]
The issues with failover in Redis sentinel is something that we've been improving via several PRs. Not sure if the same issue is still happening at the moment. Could you confirm?

**Comment by tomislater** [comment]
Yeah, I can confirm, but it is very difficult to reproduce :/

**Comment by miguelaeh** [comment]
Hi @tomislater ,
Could you share how to reproduce it please? Is it possible to reproduce it consistently?

**Comment by tomislater** [comment]
@miguelaeh 

> Could you share how to reproduce it please?

I cannot do that now, but when I come back home I will try to reproduce the issue.

> Is it possible to reproduce it consistently?

It was very difficult to reproduce it. Maybe I should try harder. I use default values btw.

I also have more than one `redis` deployed by this chart on my cluster. Maybe that's the problem? I saw recently in logs that replica from namespace `A` connected to a master from namespace `B`.

**Comment by miguelaeh** [comment]
> I saw recently in logs that replica from namespace A connected to a master from namespace B.

This definitelly should not happen. If I am not wrong, the redis replicas do not talk between them, each replica syncs with the master. And each replica sentinel is who talk with each sentinel to select the new master. Can you clarify how the traffic is flowing in your scenario please? Also, ensure the labels are properly set in the pods of both deployments, there should be a label to disting between the releases, and also, the namespaces should be included as a label in all the release resources

**Comment by tomislater** [comment]
Logs from redis in namespace `redis-A` (redis node-2):
```
1:M 03 Jan 2022 11:32:41.702 * Synchronization with replica redis-A-node-0.redis-A-headless.redis-A.svc.cluster.local:6379 succeeded
1:M 03 Jan 2022 11:43:40.134 * Replica redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local:6379 asks for synchronization
1:M 03 Jan 2022 11:43:40.134 * Partial resynchronization not accepted: Replication ID mismatch (Replica asked for '2a7a2f2585c1c39d651a29ef5e519b67ec7142ee', my replication IDs are 'aa5c90fd95d3b8414c74ed657f27d79996f761b0' and '51f08743bfd6b632c7f811a4c71a23b74200bf7f')
1:M 03 Jan 2022 11:43:40.134 * Starting BGSAVE for SYNC with target: disk
1:M 03 Jan 2022 11:43:40.135 * Background saving started by pid 2784
1:M 03 Jan 2022 11:43:40.350 # Connection with replica redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local:6379 lost.
```

And logs from sentinel node-2 (namespace `redis-A`):
```
1:X 03 Jan 2022 11:43:20.168 # -sdown slave redis-A-node-0.redis-A-headless.redis-A.svc.cluster.local:6379 redis-A-node-0.redis-A-headless.redis-A.svc.cluster.local 6379 @ A-redis redis-A-node-2.redis-A-headless.redis-A.svc.cluster.local 6379
1:X 03 Jan 2022 11:44:19.619 * +slave slave redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local:6379 redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local 6379 @ A-redis redis-A-node-2.redis-A-headless.redis-A.svc.cluster.local 6379
1:X 03 Jan 2022 11:46:57.836 # +sdown slave redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local:6379 redis-B-node-1.redis-B-headless.redis-B.svc.cluster.local 6379 @ A-redis redis-A-node-2.redis-A-headless.redis-A.svc.cluster.local 6379
1:X 03 Jan 2022 11:46:57.836 # +sdown slave redis-A-node-0.redis-A-headless.redis-A.svc.cluster.local:6379 redis-A-node-0.redis-A-headless.redis-A.svc.cluster.local 6379 @ A-redis redis-A-node-2.redis-A-headless.redis-A.svc.cluster.local 6379
```

For some reason, a redis from `redis-B` namespace asked for synchronization with redis in `redis-A` namespace.
What have I done?
I just deployed redis to `redis-B` namespace and two nodes (`node-0` and `node-2`) connected to the correct redis (in a namespace `redis-B`, but for some reason `node-1` wanted to connect to `redis-A` namespace instead of `redis-B`.
I just deleted this this pod and after that it connected to the correct namespace (`redis-B`).

Before deploying `redis-B` I deployed `redis-A`. So, maybe pod `node-1` from `redis-B` namespace got the same IP which previously had a pod in `redis-A` namespace? In this way, a redis from `redis-A` namespaces thought that pod `node-1` is a replica of `redis-A`?

How to reproduce it? Just deploy redis to two namespaces and kill pods randomly, one after another. I do not have better scenario now. I need to run more tests.

Here are my values:

```
auth:
  enabled: false
  sentinel: false

pdb:
  create: true

master:
  resources:
    requests:
      memory: 512Mi
      cpu: 100m
    limits:
      memory: 512Mi
      cpu: 1
  persistence:
    enabled: false

replica:  
  resources:
    requests:
      memory: 512Mi
      cpu: 100m
    limits:
      memory: 512Mi
      cpu: 1
  persistence:
    enabled: false

sentinel:
  enabled: true
  masterSet: B-redis
  automateClusterRecovery: true
  downAfterMilliseconds: 2000
```

I think I should create an another issue for this.

**Comment by tomislater** [comment]
In the meantime I reproduced `Unable to connect to MASTER: Invalid argument` error.

Values:

```
auth:
  enabled: false
  sentinel: false

master:
  resources:
    requests:
      memory: 128Mi
      cpu: 10m
    limits:
      memory: 128Mi
      cpu: 1
  persistence:
    enabled: false

replica:
  resources:
    requests:
      memory: 128Mi
      cpu: 10m
    limits:
      memory: 128Mi
      cpu: 1
  persistence:
    enabled: false

sentinel:
  enabled: true
  masterSet: b
  automateClusterRecovery: true
  downAfterMilliseconds: 2000
```

1. deploy the chart
2. wait for the deployment
3. kill all pods in the same time

`node-0` is up and running, but in logs you can find this:

Redis:
```
 12:37:57.13 WARN  ==> redis-headless.b.svc.cluster.local does not contain the IP of this pod: 10.50.39.189
 12:38:02.14 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
Could not connect to Redis at redis.b.svc.cluster.local:26379: Connection refused
 12:38:03.16 INFO  ==> Redis master is up now
 12:38:03.17 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel get-master-addr-by-name b
Could not connect to Redis at redis.b.svc.cluster.local:26379: Connection refused
 12:38:04.18 INFO  ==> printing REDIS_SENTINEL_INFO=(,)
1:C 03 Jan 2022 12:38:04.197 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 03 Jan 2022 12:38:04.197 # Redis version=6.2.6, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 03 Jan 2022 12:38:04.197 # Configuration loaded
1:S 03 Jan 2022 12:38:04.198 * monotonic clock: POSIX clock_gettime
1:S 03 Jan 2022 12:38:04.199 * Running mode=standalone, port=6379.
1:S 03 Jan 2022 12:38:04.199 # Server initialized
1:S 03 Jan 2022 12:38:04.199 * Ready to accept connections
1:S 03 Jan 2022 12:38:04.199 * Connecting to MASTER :0
1:S 03 Jan 2022 12:38:04.200 # Unable to connect to MASTER: Invalid argument
1:S 03 Jan 2022 12:38:05.203 * Connecting to MASTER :0
1:S 03 Jan 2022 12:38:05.203 # Unable to connect to MASTER: Invalid argument
1:S 03 Jan 2022 12:38:06.207 * Connecting to MASTER :0
1:S 03 Jan 2022 12:38:06.207 # Unable to connect to MASTER: Invalid argument
```

Sentinel:
```
 12:37:57.37 WARN  ==> redis-headless.b.svc.cluster.local does not contain the IP of this pod: 10.50.39.189
 12:38:02.38 INFO  ==> about to run the command: timeout 5 redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel get-master-addr-by-name b
Could not connect to Redis at redis.b.svc.cluster.local:26379: Connection refused
1:X 03 Jan 2022 12:38:03.512 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 03 Jan 2022 12:38:03.512 # Redis version=6.2.6, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 03 Jan 2022 12:38:03.512 # Configuration loaded
1:X 03 Jan 2022 12:38:03.513 * monotonic clock: POSIX clock_gettime
1:X 03 Jan 2022 12:38:03.514 * Running mode=sentinel, port=26379.
1:X 03 Jan 2022 12:38:03.515 # Sentinel ID is 2a09ba7abbb41ee71e79087310d75f9809c3c815
1:X 03 Jan 2022 12:38:03.515 # +monitor master b redis-node-0.redis-headless.b.svc.cluster.local 6379 quorum 2
1:X 03 Jan 2022 12:38:05.540 # +sdown slave redis-node-2.redis-headless.b.svc.cluster.local:6379 redis-node-2.redis-headless.b.svc.cluster.local 6379 @ b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:05.540 # +sdown sentinel 9fe32540b27937ed9f341b0f610a0d8df405bb63 redis-node-2.redis-headless.b.svc.cluster.local 26379 @ b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:26.596 # +sdown master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:29.438 * +sentinel sentinel 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 redis-node-1.redis-headless.b.svc.cluster.local 26379 @ b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.590 # +new-epoch 1
1:X 03 Jan 2022 12:38:49.593 # +vote-for-leader 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 1
1:X 03 Jan 2022 12:38:50.304 # +odown master b redis-node-0.redis-headless.b.svc.cluster.local 6379 #quorum 2/2
1:X 03 Jan 2022 12:38:50.304 # Next failover delay: I will not start a failover before Mon Jan  3 12:39:26 2022
1:X 03 Jan 2022 12:39:26.083 # +new-epoch 2
1:X 03 Jan 2022 12:39:26.086 # +vote-for-leader 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 2
1:X 03 Jan 2022 12:39:26.086 # Next failover delay: I will not start a failover before Mon Jan  3 12:40:02 2022
1:X 03 Jan 2022 12:40:02.352 # +new-epoch 3
1:X 03 Jan 2022 12:40:02.352 # +try-failover master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:40:02.355 # +vote-for-leader 2a09ba7abbb41ee71e79087310d75f9809c3c815 3
```

In `node-1` readiness probe fails (but sentinel works):
Redis:
```
 12:44:25.69 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
 12:44:25.69 INFO  ==> sleeping, waiting for Redis master to come up
 12:44:26.70 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
 12:44:26.71 INFO  ==> sleeping, waiting for Redis master to come up
 12:44:27.71 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
 12:44:27.71 INFO  ==> sleeping, waiting for Redis master to come up
 12:44:28.72 INFO  ==> Trying to manually failover
 12:44:28.72 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel failover b
 12:44:28.73 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
 12:44:28.74 INFO  ==> sleeping, waiting for Redis master to come up
```

Sentinel:
```
 12:38:22.33 WARN  ==> redis-headless.b.svc.cluster.local does not contain the IP of this pod: 10.50.32.68
 12:38:27.34 INFO  ==> about to run the command: timeout 5 redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel get-master-addr-by-name b
redis-node-0.redis-headless.b.svc.cluster.local
6379
 12:38:27.35 INFO  ==> about to run the command: timeout 5 redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel get-master-addr-by-name b
 12:38:27.36 INFO  ==> printing REDIS_SENTINEL_INFO=(redis-node-0.redis-headless.b.svc.cluster.local,6379)
1:X 03 Jan 2022 12:38:27.413 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 03 Jan 2022 12:38:27.413 # Redis version=6.2.6, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 03 Jan 2022 12:38:27.413 # Configuration loaded
1:X 03 Jan 2022 12:38:27.414 * monotonic clock: POSIX clock_gettime
1:X 03 Jan 2022 12:38:27.415 * Running mode=sentinel, port=26379.
1:X 03 Jan 2022 12:38:27.415 # Sentinel ID is 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2
1:X 03 Jan 2022 12:38:27.415 # +monitor master b redis-node-0.redis-headless.b.svc.cluster.local 6379 quorum 2
1:X 03 Jan 2022 12:38:29.437 # +sdown slave redis-node-1.redis-headless.b.svc.cluster.local:6379 redis-node-1.redis-headless.b.svc.cluster.local 6379 @ b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.486 # +sdown master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.564 # +odown master b redis-node-0.redis-headless.b.svc.cluster.local 6379 #quorum 2/2
1:X 03 Jan 2022 12:38:49.564 # +new-epoch 1
1:X 03 Jan 2022 12:38:49.564 # +try-failover master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.575 # +vote-for-leader 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 1
1:X 03 Jan 2022 12:38:49.593 # 2a09ba7abbb41ee71e79087310d75f9809c3c815 voted for 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 1
1:X 03 Jan 2022 12:38:49.672 # +elected-leader master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.672 # +failover-state-select-slave master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.727 # -failover-abort-no-good-slave master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:38:49.793 # Next failover delay: I will not start a failover before Mon Jan  3 12:39:25 2022
1:X 03 Jan 2022 12:39:26.061 # +new-epoch 2
1:X 03 Jan 2022 12:39:26.061 # +try-failover master b redis-node-0.redis-headless.b.svc.cluster.local 6379
1:X 03 Jan 2022 12:39:26.068 # +vote-for-leader 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2 2
```

```
➜ kubectl get pods --namespace b --context dev
NAME           READY   STATUS    RESTARTS   AGE
redis-node-0   2/2     Running   0          10m
redis-node-1   1/2     Running   6          9m55s
```

**Comment by tomislater** [comment]
And here we are. I killed `redis-node-0` (`b` namespace) and it connected to `a` namespace.

Logs from `redis-node-0` (`b` namespace).
```
1:S 03 Jan 2022 12:52:24.505 # Unable to connect to MASTER: Invalid argument
1:S 03 Jan 2022 12:52:25.509 * Connecting to MASTER :0
1:S 03 Jan 2022 12:52:25.509 # Unable to connect to MASTER: Invalid argument
1:S 03 Jan 2022 12:52:25.791 * Connecting to MASTER redis-node-2.redis-headless.a.svc.cluster.local:6379
1:S 03 Jan 2022 12:52:25.792 * MASTER <-> REPLICA sync started
1:S 03 Jan 2022 12:52:25.792 * REPLICAOF redis-node-2.redis-headless.a.svc.cluster.local:6379 enabled (user request from 'id=5 addr=10.50.36.117:50522 laddr=10.50.32.68:6379 fd=11 name=sentinel-33535e4e-cmd age=18 idle=0 flags=x db=0 sub=0 psub=0 multi=4 qbuf=235 qbuf-free=40719 argv-mem=4 obl=45 oll=0 omem=0 tot-mem=61468 events=r cmd=exec user=default redir=-1')
1:S 03 Jan 2022 12:52:25.792 * Non blocking connect for SYNC fired the event.
1:S 03 Jan 2022 12:52:25.792 * Master replied to PING, replication can continue...
1:S 03 Jan 2022 12:52:25.793 * Partial resynchronization not possible (no cached master)
1:S 03 Jan 2022 12:52:25.793 * Full resync from master: eed3b9678ee61e1e5c5bc105baacc18535bc6dae:368349
1:S 03 Jan 2022 12:52:25.873 * MASTER <-> REPLICA sync: receiving 178 bytes from master to disk
1:S 03 Jan 2022 12:52:25.873 * MASTER <-> REPLICA sync: Flushing old data
1:S 03 Jan 2022 12:52:25.874 * MASTER <-> REPLICA sync: Loading DB in memory
```

Logs from `redis-node-2` (`a` namespace).
```
494:C 03 Jan 2022 12:36:12.383 * RDB: 0 MB of memory used by copy-on-write
1:M 03 Jan 2022 12:36:12.383 * Background saving terminated with success
1:M 03 Jan 2022 12:36:12.383 * Synchronization with replica redis-node-0.redis-headless.a.svc.cluster.local:6379 succeeded
1:M 03 Jan 2022 12:52:25.793 * Replica redis-node-0.redis-headless.b.svc.cluster.local:6379 asks for synchronization
1:M 03 Jan 2022 12:52:25.793 * Full resync requested by replica redis-node-0.redis-headless.b.svc.cluster.local:6379
1:M 03 Jan 2022 12:52:25.793 * Starting BGSAVE for SYNC with target: disk
1:M 03 Jan 2022 12:52:25.793 * Background saving started by pid 4372
4372:C 03 Jan 2022 12:52:25.795 * DB saved on disk
4372:C 03 Jan 2022 12:52:25.796 * RDB: 0 MB of memory used by copy-on-write
1:M 03 Jan 2022 12:52:25.873 * Background saving terminated with success
1:M 03 Jan 2022 12:52:25.873 * Synchronization with replica redis-node-0.redis-headless.b.svc.cluster.local:6379 succeeded
```

It is very hard to reproduce. Just try to kill a pod every `xx` seconds.

**Comment by tomislater** [comment]
I think that there is something wrong with this line: https://github.com/bitnami/charts/blob/master/bitnami/redis/templates/scripts-configmap.yaml#L119

When I kill all pods at the same time the first one (`node-0`) which should be master (because there is no any other pod) "thinks" that it is a replica.

`validate_quorum` returns `0`:
```
 15:58:34.25 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel master b
Could not connect to Redis at redis.b.svc.cluster.local:26379: Connection refused
```

Then in logs you can find:
```
 15:58:35.29 INFO  ==> Redis master is up now
```
which is misleading. And after that we got:
```
 15:58:35.29 INFO  ==> about to run the command: redis-cli -h redis.b.svc.cluster.local -p 26379 sentinel get-master-addr-by-name b
Could not connect to Redis at redis.b.svc.cluster.local:26379: Connection refused
 15:58:36.32 INFO  ==> printing REDIS_SENTINEL_INFO=(,)
```
Another silent error.

And because `node-0` thinks that it is a slave, it tries to connect to not existing address (`REDIS_SENTINEL_INFO=(,)`): https://github.com/bitnami/charts/blob/master/bitnami/redis/templates/scripts-configmap.yaml#L185

**Comment by tomislater** [comment]
Changing
```
if [[ -z "$(getent ahosts "$HEADLESS_SERVICE" | grep -v "^${myip}")" ]]; then
```

into

```
redis-cli -h $REDIS_SERVICE -p $SENTINEL_SERVICE_PORT sentinel get-master-addr-by-name b
retVal=$?
if [[ $retVal -ne 0 ]]; then
```

fixes the issue with `Unable to connect to MASTER: Invalid argument`. I explicitly checking if there is a master already. If not then I assume that I am the master.

**Comment by miguelaeh** [comment]
Hi @tomislater ,

About the quorum issue, it would be great if you could create the PR with the suggested code.

About the issue of re-using IPs between namespaces, I think it is something derivated from the fact that Redis does not support being configured using hostnames. Since we need to configure it using the pod IPs you could try adding network policies to prevent pods of different namespaces to connect. I think the IP reuse between namespaces is probably something that can be configured, but I am not sure how right now, maybe in the CNI configuration.

**Comment by tomislater** [comment]
Hi @miguelaeh,

I am going to create PR with my changes today.

**Comment by tomislater** [comment]
> About the issue of re-using IPs between namespaces, I think it is something derivated from the fact that Redis does not support being configured using hostnames.

Hmm, we use hostnames by default in this chart (https://github.com/bitnami/charts/blob/master/bitnami/redis/templates/scripts-configmap.yaml#L365). We don't use IPs anymore. That's why it is strange to me. Looks like `sentinel` uses IPs internally despite this?

**Comment by tomislater** [comment]
@miguelaeh here is my fix: https://github.com/bitnami/charts/pull/8563

**Comment by tomislater** [comment]
Sentinel (which resides in `a` namespace)  added a redis from `b` namespace as a replica.
```
cat cd /opt/bitnami/redis-sentinel/etc/sentinel.conf 

dir "/tmp"
port 26379
sentinel monitor a redis-node-1.redis-headless.a.svc.cluster.local 6379 2
sentinel down-after-milliseconds a 2000
sentinel failover-timeout a 18000

# User-supplied sentinel configuration:
# End of sentinel configuration
sentinel myid 9fe32540b27937ed9f341b0f610a0d8df405bb63

sentinel known-sentinel a redis-node-0.redis-headless.a.svc.cluster.local 26379 2a09ba7abbb41ee71e79087310d75f9809c3c815

sentinel known-replica a redis-node-2.redis-headless.b.svc.cluster.local 6379

sentinel known-sentinel a redis-node-1.redis-headless.a.svc.cluster.local 26379 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2

sentinel known-replica a redis-node-0.redis-headless.a.svc.cluster.local 6379

sentinel announce-hostnames yes
sentinel resolve-hostnames yes
sentinel announce-port 26379
sentinel announce-ip "redis-node-2.redis-headless.a.svc.cluster.local"
# Generated by CONFIG REWRITE
protected-mode no
user default on nopass ~* &* +@all
sentinel config-epoch a 5
sentinel leader-epoch a 5
sentinel current-epoch 5

sentinel known-replica a redis-node-2.redis-headless.a.svc.cluster.local 6379
```

This line:
```
sentinel known-replica a redis-node-2.redis-headless.b.svc.cluster.local 6379
```

This happen regularly if you use spot instances (AWS). You can just kill pods also. I need to debug this problem deeper.

**Comment by tomislater** [comment]
In [this chart](https://github.com/DandyDeveloper/charts/tree/master/charts/redis-ha) they are using a `Service` for every pod:

<img width="1621" alt="Screen Shot 2022-01-05 at 13 00 56" src="https://user-images.githubusercontent.com/1127164/148214577-3d9de187-a595-4bf8-9da9-e6194f6825c7.png">

Which makes a lot of sense because these IPs do not change. Here we are using headless `Service` and sentinel can connect to the wrong redis/sentinel.

**Comment by miguelaeh** [comment]
Hi @tomislater ,
> In this chart they are using a Service for every pod:

That is something we would like to avoid if possible. There are cases where it is totally necessary, you can check our redis-cluster or cassandra charts. In those charts, that approach is used for external access using LoadBalancer services to each pod.


I think you could configure the Network Policy included in the chart to avoid that connection to happen. Check [this](https://github.com/bitnami/charts/blob/master/bitnami/redis/templates/networkpolicy.yaml). You can configure custom `namespaceSelectors` in your values.yaml that will applied there. More info [here](https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors)
.

**Comment by tomislater** [comment]
Hi @miguelaeh

thanks for suggestions!

I checked yesterday `NetworkPolicy`, but we don't want to install another tool (calico) for such simple case. Additionally, using `NetworkPolicy` is not a good approach in my view. Sentinels still wanted to connect to wrong redis/sentinel, but such connections would be prevented.

Thus, we chose to use `dandydev/redis-ha` chart. Thanks!
